src/misc/unicode.cpp

```diff
@@ -451,712 +653,878 @@ static bool wide_to_code_page(const std::vector<uint16_t> &str_in,
-        return false;
-    }
-
-    construct_aliases(code_page);
-    return true;
+	if (mappings_normalized_by_codepage.count(code_page)) {
+		return true; // code page already prepared
+	}
+
+	if (!config_mappings.count(code_page) || !construct_mapping(code_page)) {
+		// Unsupported code page or error
+		mappings_normalized_by_codepage.erase(code_page);
+		mappings_decomposed_by_codepage.erase(code_page);
+		mappings_reverse_by_codepage.erase(code_page);
+		return false;
+	}
+
+	construct_aliases(code_page);
+	return true;
 }
 
 static void load_config_if_needed()
 {
-    // If this is the first time we are requested to prepare the code page,
-    // load the top-level configuration and fallback 7-bit ASCII mapping
-
-    static bool config_loaded = false;
-    if (!config_loaded) {
-        const auto path_root = GetResourcePath(dir_name_mapping);
-        import_mapping_ascii(path_root);
-        import_config_main(path_root);
-        config_loaded = true;
-    }
+	// If this is the first time we are requested to prepare the code page,
+	// load the top-level configuration and fallback 7-bit ASCII mapping
+
+	static bool config_loaded = false;
+	if (!config_loaded) {
+		const auto path_root = GetResourcePath(dir_name_mapping);
+		import_decomposition(path_root);
+		import_mapping_ascii(path_root);
+		import_config_main(path_root);
+		config_loaded = true;
+	}
+}
+
+static uint16_t get_default_code_page()
+{
+	constexpr uint16_t default_code_page = 437; // United States
+
+	if (!prepare_code_page(default_code_page)) {
+		warn_default_code_page();
+		return 0;
+	}
+
+	return default_code_page;
+}
+
+static uint16_t get_custom_code_page(const uint16_t custom_code_page)
+{
+	if (custom_code_page == 0) {
+		return 0;
+	}
+
+	const uint16_t code_page = deduplicate_code_page(custom_code_page);
+	if (!prepare_code_page(code_page)) {
+		return get_default_code_page();
+	}
+
+	return code_page;
 }
 
 // ***************************************************************************
 // External interface
 // ***************************************************************************
 
-uint16_t UTF8_GetCodePage()
+uint16_t get_utf8_code_page()
 {
-    const uint16_t cp_default = 437; // United States
-    if (!IS_EGAVGA_ARCH)
-        // Below EGA it wasn't possible to change character set
-        return cp_default;
-
-    load_config_if_needed();
-    const uint16_t cp = deduplicate_code_page(dos.loaded_codepage);
-
-    // For unsupported code pages, revert to default one
-    if (prepare_code_page(cp))
-        return cp;
-    else
-        return cp_default;
+	if (!IS_EGAVGA_ARCH) {
+		// Below EGA it wasn't possible to change the character set
+		return get_default_code_page();
+	}
+
+	load_config_if_needed();
+	const uint16_t code_page = deduplicate_code_page(dos.loaded_codepage);
+
+	// For unsupported code pages revert to default one
+	if (prepare_code_page(code_page)) {
+		return code_page;
+	} else {
+		return get_default_code_page();
+	}
```

Comment:
The `else ... return` can be flatten away to just `return`.

Reply:
Done.